libpromises/evalfunction.c

```diff
@@ -1695,6 +1701,265 @@ static FnCallResult FnCallSplayClass(EvalContext *ctx,
+        }
+        else if (0 == strcmp(key, "url.user-agent"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting user agent string to %s", fp->name, value);
+            curl_easy_setopt(curl, CURLOPT_USERAGENT, value);
+        }
+        else if (0 == strcmp(key, "url.max_content"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting max contents to %ld", fp->name, IntFromString(value));
+            max_content = IntFromString(value);
+        }
+        else if (0 == strcmp(key, "url.max_headers"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting max headers to %ld", fp->name, IntFromString(value));
+            max_headers = IntFromString(value);
+        }
+        else
+        {
+            Log(LOG_LEVEL_INFO, "%s: unknown option %s", fp->name, key);
+        }
+    }
+
+    struct _curl_userdata data = { fp, "content", max_content, content };
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &data);
+
+    struct _curl_userdata header_data = { fp, "headers", max_headers, headers };
+    curl_easy_setopt(curl, CURLOPT_HEADERDATA, &header_data);
+
+    JsonElement *options_headers = JsonObjectGetAsArray(options, "url.headers");
+    struct curl_slist *header_list = NULL;
+
+    if (NULL != options_headers)
+    {
+        iter = JsonIteratorInit(options_headers);
+        while ((e = JsonIteratorNextValueByType(&iter, JSON_ELEMENT_TYPE_PRIMITIVE, true)))
+        {
+            header_list = curl_slist_append(header_list, JsonPrimitiveGetAsString(e));
+        }
+    }
+
+    if (NULL != header_list)
+    {
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, header_list);
+    }
+
+    JsonElement *result = JsonObjectCreate(10);
+    CURLcode res = curl_easy_perform(curl);
+    if (NULL != header_list)
+    {
+        curl_slist_free_all(header_list);
+        header_list = NULL;
+    }
+
+    long returncode = 0;
+    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &returncode);
+    JsonObjectAppendInteger(result, "returncode",  returncode);
+
+    curl_easy_cleanup(curl);
+
+    JsonObjectAppendInteger(result, "rc",  0+res);
+
+    bool success = (CURLE_OK == res);
+    JsonObjectAppendBool(result, "success", success);
+
+    if (!success)
+    {
+        JsonObjectAppendString(result, "error_message", curl_easy_strerror(res));
+    }
+
+
+
+    BufferTrimToMaxLength(content, max_content);
+    JsonObjectAppendString(result, "content",  BufferData(content));
+    BufferDestroy(content);
+
+    BufferTrimToMaxLength(headers, max_headers);
+    JsonObjectAppendString(result, "headers",  BufferData(headers));
+    BufferDestroy(headers);
+
+    JsonObjectAppendObject(CURL_CACHE, StringWriterData(cache_w), JsonCopy(result));
+    WriterClose(cache_w);
+
+    JsonDestroy(options);
+    return (FnCallResult) { FNCALL_SUCCESS, (Rval) { result, RVAL_TYPE_CONTAINER } };
+#else
+    Log(LOG_LEVEL_INFORM,
+        "%s: libcurl integration is not compiled into CFEngine, sorry", fp->name);
+    return FnFailure();
```

Comment:
FYI, the correct log level is ``LOG_LEVEL_INFO``, but in any case, I think ``LOG_LEVEL_ERR`` is more appropriate here, because the user tried to do something that can't be done.